pkg/caveats/eval.go

```diff
@@ -106,45 +102,39 @@ func EvaluateCaveatWithConfig(caveat *CompiledCaveat, contextValues map[string]a
 		return nil, err
 	}
 
-	pvars, err := cel.PartialVars(contextValues)
+	// Mark any unspecified variables as unknown, to ensure that partial application
+	// will result in producing a type of Unknown.
+	activation, err := env.PartialVars(contextValues)
 	if err != nil {
 		return nil, err
 	}
 
-	val, details, err := prg.Eval(pvars)
-	if err != nil {
-		// From program.go:
-		// *  `val`, `details`, `nil` - Successful evaluation of a non-error result.
-		// *  `val`, `details`, `err` - Successful evaluation to an error result.
-		// *  `nil`, `details`, `err` - Unsuccessful evaluation.
-		//
-		// NOTE: This is done in this hacky way right now because CEL does not have
-		// well-typed errors. We should change to a better way to detect partial
-		// eval if/when CEL adds properly wrapped errors.
-		// See: https://github.com/google/cel-go/issues/25
-		if val != nil && strings.Contains(err.Error(), "no such attribute") {
-			found := noSuchAttributeErrMessage.FindStringSubmatch(err.Error())
-			if found != nil {
-				return &CaveatResult{
-					val:             val,
-					details:         details,
-					parentCaveat:    caveat,
-					contextValues:   contextValues,
-					missingVarNames: strings.Split(found[2], " "),
-					isPartial:       true,
-				}, nil
+	val, details, err := prg.Eval(activation)
+
+	// If the value produced has Unknown type, then it means required context was missing.
+	if types.IsUnknown(val) {
+		unknownVal := val.(*types.Unknown)
+		missingVarNames := make([]string, 0, len(contextValues))
+		for _, id := range unknownVal.IDs() {
+			trails, ok := unknownVal.GetAttributeTrails(id)
+			if ok {
+				for _, attributeTrail := range trails {
+					missingVarNames = append(missingVarNames, attributeTrail.Variable())
```

Comment:
Consider using `AttributeTrail.String()` in the event that you have more specific missing attributes in the future.

Reply:
Changed